toolchain/check/testdata/builtin_conversions/no_prelude/convert_facet_value_as_type_knows_original_type.carbon

```diff
@@ -0,0 +1,573 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// AUTOUPDATE
+// TIP: To test this file alone, run:
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/check/testdata/builtin_conversions/no_prelude/convert_facet_value_as_type_knows_original_type.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/builtin_conversions/no_prelude/convert_facet_value_as_type_knows_original_type.carbon
+
+// --- core.carbon
+
+package Core;
+
+interface As(Dest:! type) {
+  fn Convert[self: Self]() -> Dest;
+}
+
+interface ImplicitAs(Dest:! type) {
+  fn Convert[self: Self]() -> Dest;
+}
+
+// --- explicit_as_type.carbon
+
+library "[[@TEST_NAME]]";
+
+import Core;
+
+interface Eats {}
+interface Animal {}
+
+class Goat {}
+impl Goat as Animal {}
+impl Goat as Eats {}
+
+fn Feed(e:! Eats) {}
+
+fn F() {
+  Feed((Goat as Animal) as type);
+}
+
+// --- facet_type_in_type_position.carbon
+
+library "[[@TEST_NAME]]";
+
+import Core;
+
+interface Eats {
+  fn Eat();
+}
+interface Animal {}
+
+class Goat {
+  fn Bleet() {}
+
+  impl as Animal {}
+  extend impl as Eats {
+    fn Eat() {}
+  }
+}
+
+fn F() {
+  // `Goat as Animal` in the type position retains/recovers the original type
+  // Goat, so member lookup can see more than just `Animal`.
+
+  let x: Goat as Animal = {} as Goat;
+  x.Bleet();
+  x.Eat();
+
+  (({} as Goat) as (Goat as Animal)).Bleet();
+  (({} as Goat) as (Goat as Animal)).Eat();
+}
+
+// CHECK:STDOUT: --- core.carbon
```

Comment:
Do you think the SemIR output is useful in this test, or should we disable it? It looks like the main point of the test is to ensure the above code doesn't produce diagnostics.

Reply:
This was a semi-rare case where I did actually want to look at the semir after the test was written to make sure it's doing the right thing - to confirm that an `as type` was happening for the call to `Eat()` and `Bleet` was going to the right place. So maybe we can keep it for this one.